### PR TITLE
New test for checking websocket client-server interaction

### PR DESCRIPTION
**Pull request by laura-eprosima**
The PR #43 need to be merged for this new test to pass.

### CODE REVIEW

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss/129/)

File: packages/websocket-test/integration/websocket__services.cpp [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+        , _promise(std::move(promise))
+        , _req_type(req_type)
+        , _response(response)
+    {
+        YAML::Node configuration = YAML::LoadFile(WEBSOCKET__SERVICES__TEST_CONFIG);
+        // Gets the port from the websocket client defined on the yaml
+        port = configuration["systems"]["ws_client"]["port"].as<int>();
+        if (configuration["security"] && configuration["security"].as<std::string>() == "none")
+        {
+            std::cout << "Security Disabled -> Using TCP" << std::endl;
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:66]:
```suggestion
            std::cout << "[soss-websocket-test] Security Disabled -> Using TCP" << std::endl;
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+            security = false;
+            _thread = new std::thread(std::bind(&ServerTest::configure_tcp_server, this));
+        }
+        else
+        {
+            std::cout << "Security Enabled -> Using TLS" << std::endl;
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:72]:
```suggestion
            std::cout << "[soss-websocket-test] Security Enabled -> Using TLS" << std::endl;
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+
+        std::vector<std::string> checked_paths;
+
+        checked_paths.clear();
+        std::string cert_file = "certs/websocket_test.crt";
+        const std::string cert_file_path = ca_search.find_file(cert_file, "", &checked_paths);
+
+        checked_paths.clear();
+        std::string key_file = "certs/websocket_test.key";
+        const std::string key_file_path = ca_search.find_file(key_file, "", &checked_paths);
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:97]: `checked_paths` not used, `soss::Search::find_file` already defaults third parameter to `nullptr` if not used

Comment (jamoralp) [reply packages/websocket-test/integration/websocket__services.cpp:97]: Also second parameter is defaulted to `""`. I suggest just leaving it as `ca_search.find_file(key_file)`

File: packages/websocket-test/integration/websocket__services.cpp [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+
+        // Looks for the security certificate and key
+        soss::Search ca_search = soss::Search("websocket")
+                .relative_to_config()
+                .relative_to_home();
+
+        std::vector<std::string> checked_paths;
+
+        checked_paths.clear();
+        std::string cert_file = "certs/websocket_test.crt";
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:92]:
```suggestion
        const std::string cert_file = "certs/websocket_test.crt";
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+        const std::string cert_file_path = ca_search.find_file(cert_file, "", &checked_paths);
+
+        checked_paths.clear();
+        std::string key_file = "certs/websocket_test.key";
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:96]:
```suggestion
        const std::string key_file = "certs/websocket_test.key";
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+        const std::string key_file_path = ca_search.find_file(key_file, "", &checked_paths);
+
+        _context->use_certificate_file(cert_file_path, SslContext::pem, ec);
+        if (ec)
+        {
+            std::cerr << "[soss::websocket::Server] Failed to load certificate file ["
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:102]:
```suggestion
            std::cerr << "[soss-websocket-test][websocket::Server] Failed to load certificate file ["
```

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss/135/)

File: packages/websocket-test/integration/websocket__services.cpp [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+                      << cert_file << "]: " << ec.message() << std::endl;
+        }
+
+        _context->use_rsa_private_key_file(key_file_path, SslContext::pem, ec);
+        if (ec)
+        {
+            std::cerr << "[soss::websocket::Server] Failed to load private key file ["
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:109]:
```suggestion
            std::cerr << "[soss-websocket-test][websocket::Server] Failed to load private key file ["
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+            });
+        _tls_server->set_close_handler(
+            [&](websocketpp::connection_hdl handle)
+            {
+                _mutex->lock();
+                const auto connection = _tls_server->get_con_from_hdl(handle);
+                _open_tls_connections.erase(connection);
+                _mutex->unlock();
+            });
+        std::cout << "Setting port " << port << std::endl;
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:153]:
```suggestion
        std::cout << "[soss-websocket-test][websocket::Server] Listening to port: " << port << std::endl;
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+            });
+        _tcp_server->set_close_handler(
+            [&](websocketpp::connection_hdl handle)
+            {
+                _mutex->lock();
+                const auto connection = _tcp_server->get_con_from_hdl(handle);
+                _open_tcp_connections.erase(connection);
+                _mutex->unlock();
+            });
+        std::cout << "Setting port " << port << std::endl;
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:207]:
```suggestion
        std::cout << "[soss-websocket-test][websocket::Server] Listening to port: " << port << std::endl;
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+    }
+
+    // Manages the messages received on the TLS Server
+    void on_tls_message(
+            websocketpp::connection_hdl hdl,
+            TlsServer::message_ptr msg)
+    {
+        // Sets the server_promise value to notify that the request has arrived correctly
+        const auto json_msg = nlohmann::json::parse(msg->get_payload());
+        std::cout << "Msg Received on Test Websocket Server: " << msg->get_payload() << std::endl;
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:220]:
```suggestion
        std::cout << "[soss-websocket-test][websocket::Server] Msg Received: '" << msg->get_payload() << "'" << std::endl;
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+        const auto req_it = json_msg.find("args");
+        if (req_it != json_msg.end())
+        {
+            auto d_data = soss::json::convert(_req_type, req_it.value());
+            _promise->set_value(d_data);
+
+            // Sends the response
+            auto connection = _tls_server->get_con_from_hdl(hdl);
+            std::string response_msg =
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:229]:
```suggestion
            const std::string response_msg =
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+                    "{\"op\":\"service_response\",\"type\":\"Data_Response\",\"service\":\"client_request\",\"id\":\"1\", \"values\":"
+                    + soss::json::convert(_response).dump() + ", \"result\":\"true\"}";
+            std::cout << "Sending Response: " << response_msg << std::endl;
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:232]:
```suggestion
            std::cout << "[soss-websocket-test][websocket::Server] Sending Response: '" << response_msg << "'" << std::endl;
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+        }
+    }
+
+    void on_tcp_message(
+            websocketpp::connection_hdl hdl,
+            TcpServer::message_ptr msg)
+    {
+        // Sets the server_promise value to notify that the request has arrived correctly
+        const auto json_msg = nlohmann::json::parse(msg->get_payload());
+        std::cout << "Msg Received on Test Websocket Server: " << msg->get_payload() << std::endl;
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:243]:
```suggestion
        std::cout << "[soss-websocket-test][websocket::Server] Msg Received: '" << msg->get_payload() << "'" << std::endl;
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+        const auto req_it = json_msg.find("args");
+        if (req_it != json_msg.end())
+        {
+            auto d_data = soss::json::convert(_req_type, req_it.value());
+            _promise->set_value(d_data);
+
+            // Sends the response
+            auto connection = _tcp_server->get_con_from_hdl(hdl);
+            std::string response_msg =
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:252]:
```suggestion
            const std::string response_msg =
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 15]
```diff
@@ -0,0 +1,400 @@
...
+                    "{\"op\":\"service_response\",\"type\":\"Data_Response\",\"service\":\"client_request\",\"id\":\"1\", \"values\":"
+                    + soss::json::convert(_response).dump() + ", \"result\":\"true\"}";
+            std::cout << "Sending Response: " << response_msg << std::endl;
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:255]:
```suggestion
            std::cout << "[soss-websocket-test][websocket::Server] Sending Response: '" << response_msg << "'" << std::endl;
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+        if (security)
+        {
+            _mutex->lock();
+            // Close all the connections before stopping the server
+            for (auto connection : _open_tls_connections)
+            {
+                if (connection->get_state() != websocketpp::session::state::closed)
+                {
+                    std::cout << "Closing an unclosed connection" << std::endl;
+                    try {
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:271]:
```suggestion
                    try
                    {
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+    {
+        if (security)
+        {
+            _mutex->lock();
+            // Close all the connections before stopping the server
+            for (auto connection : _open_tls_connections)
+            {
+                if (connection->get_state() != websocketpp::session::state::closed)
+                {
+                    std::cout << "Closing an unclosed connection" << std::endl;
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:270]:
```suggestion
                    std::cout << "[soss-websocket-test] Closing an unclosed connection" << std::endl;
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 18]
```diff
@@ -0,0 +1,400 @@
...
+                    try {
+                        connection->close(websocketpp::close::status::normal, "shutdown");
+                    }
+                    catch (websocketpp::exception e)
+                    {
+                        std::cout << "Exception " << e.what() << std::endl;
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:276]:
```suggestion
                        std::cout << "[soss-websocket-test] Exception " << e.what() << std::endl;
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 19]
```diff
@@ -0,0 +1,400 @@
...
+                        std::cout << "connection status " <<
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:277]:
```suggestion
                        std::cout << "[soss-websocket-test] Connection status: " <<
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 20]
```diff
@@ -0,0 +1,400 @@
...
+    // Send the request using the Mock Client created in SOSS
+    std::shared_future<xtypes::DynamicData> response_future =
+            soss::mock::request("client_request", request_msg);
+
+    // First Route: Test Mock Request -> SOSS Mock Client -> SOSS Websocket Client (server) -> Test Websocket Server
+    {
+        // Asures that the request is correctly received on the websocket server
+        REQUIRE(std::future_status::ready == server_future.wait_for(2s));
+        auto request = std::move(server_future.get());
+        std::cout << "Request " << request << std::endl;
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:367]:
```suggestion
        std::cout << "[soss-websocket-test] Service request: " << request << std::endl;
```

File: packages/websocket-test/integration/websocket__services.cpp [hunk 21]
```diff
@@ -0,0 +1,400 @@
...
+        CHECK(request == request_msg);
+    }
+
+    // Second Route: Test Websocket Server -> SOSS Websocket Client (server) -> SOSS Mock Client
+    {
+        REQUIRE(std::future_status::ready == response_future.wait_for(2s));
+        xtypes::DynamicData response = std::move(response_future.get());
+        std::cout << "Response " << response << std::endl;
```
Comment (jamoralp) [inline packages/websocket-test/integration/websocket__services.cpp:375]:
```suggestion
        std::cout << "[soss-websocket-test] Service response " << response << std::endl;
```

**Review by jamoralp — CHANGES_REQUESTED** [review]

**Review by jamoralp — APPROVED** [review]

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/soss/137/)

**Pull request merged**